Initial commit v0.1

Ashrjz opened the pull request:
Flask-based REST API for v0.1

The error handling can be done better but I did not know which fields in the professional data are optional, so I did not explicitly handle those errors, let me know if you want it done.

Let me know the changes you want.

Ashrjz committed:
Initial commit v0.1

Flask based REST API for v0.1

ab-emptycup commented:
Great! I'll review and get back to you.

On Wed, 8 Feb, 2023, 18:14 Ashish Raj, ***@***.***> wrote:

> Flask-based REST API for v0.1
>
> The error handling can be done better but I did not know which fields in
> the professional data are optional, so I did not explicitly handle those
> errors, let me know if you want it done.
>
> Let me know the changes you want.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/EmptyCupHQ/portfolio/pull/12
> Commit Summary
>
>    - d7215c2
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/commits/d7215c2844ddb90ab8882635fcfff447a80a4260>
>    Initial commit v0.1
>
> File Changes
>
> (8 files <https://github.com/EmptyCupHQ/portfolio/pull/12/files>)
>
>    - *M* .gitignore
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/files#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947>
>    (2)
>    - *A* requirements.txt
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552>
>    (12)
>    - *A* src/api/__init__.py
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/files#diff-52786a72fe30975b48dd7fe8f0bae173223b8d1d1b93623cce5af0ba6e24d447>
>    (1)
>    - *D* src/api/api.py
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/files#diff-608d096921cb2e41aa9dc8efa453ce6aa3a244ce98a66243edaa49337f2a54ae>
>    (1)
>    - *A* src/api/app.py
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/files#diff-dc5c3ef4b3c46ab4dd70d6066f139050bb3d83acd161cb30b59b7552741b3269>
>    (19)
>    - *A* src/api/db.json
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/files#diff-90174ab3e27dacfd62ecb1483cfec8e24b8ec667cf069f857938162db2a5f4f8>
>    (38)
>    - *A* src/api/main.py
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/files#diff-0e9f286abebcdf99cfb2d7c5a564b04353c86ff462af7345f49c3038a3c7624b>
>    (20)
>    - *A* src/api/pros.py
>    <https://github.com/EmptyCupHQ/portfolio/pull/12/files#diff-61c9ccca30089d569fa49941ebec01e1115b1062d5541519b4d825001f58e5da>
>    (148)
>
> Patch Links:
>
>    - https://github.com/EmptyCupHQ/portfolio/pull/12.patch
>    - https://github.com/EmptyCupHQ/portfolio/pull/12.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/EmptyCupHQ/portfolio/pull/12>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT3ZNBFAGDAYF5MJXUKLU7LWWOIMBANCNFSM6AAAAAAUVGYGXA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

ab-emptycup commented:
Hi Ashish,

Good work with the PR. Here's some feedback on the PR:

- Can we drop the *__init__.py*? *__init__.py* typically means that the code in the folder can be imported like a package. Since we don't need to import the api, may be we can delete that file. Did you include it for some other reason? May be to get the api to run properly from the terminal?
- _db.json_ is a data file. Generally it is a good convention to not include data files in the code commit. You can add that into the _.gitignore_ file. But, you should also make sure that your code doesn't give an error if _db.json_ is not available on the first run. If the file is not present, it should create it.
- I think we should be able to merge _app.py_ and _main.py_. The general convention for flask is that the _app.py_ file is the one that contains the driver / entry point. It is invoked by default on _flask run_. _main.py_ is generally used programs that are run from the command line. But that is just a convention. What was your thought process for separating _app.py_ and _main.py_?
- Another matter of convention is naming _app.py_. Typically "app" is used for complete backends that serve static files also. In our case, all the static files (eg: frontend html pages, images, logo, favicon, css sheets, js, fonts etc) will be served separately from a static hosting provider like netlify. The flask based backend will be plain REST api. Hence, it is recommended to follow the convention of using the name _api_ instead of _app_ or _main_ everywhere. Running the dev server may have to change into `flask run --app api`. The App variable inside _api.py_ can also be named _api_, so that the routes can be defined as _@api.route_.
- The common convention in python is to use `snake_casing` instead of `camelCasing` for naming variables. For eg: `listProfessionals` becomes `list_professionals`. Further I'd recommend using `list_pros` for cleanliness, brevity and maintaining consistency with the _pro_ keyword used in the api routes.

There are a few more pointers on the code. Let's do a quick code walk through during the call today. Good work overall. Let's keep at it to build some momentum. Kudos.

ashishRaj-0 commented:
So, for handling HTTP errors, I was thinking to do
```Python
@app.errorhandler(HTTPException)
def http_error(error):
    """
    Handles all HTTP errors
    """
    return jsonify({
        'code': error.code,
        'message': error.name,
        'description': error.description
    }), error.code
```
This looks cleaner, as most errors will have the same structure. And while raising errors using `abort(error_code, description)`.

I can do the other way as well handling each error separately. Something like
```python
@app.errorhandler(500)
def internal_server_error(error):
    """
    Handles 500 errors
    """
    return jsonify({'error': str(error)}), 500
```
for each error type.

Let me know.

ab-emptycup commented:
You can go with option 1. Seems more concise. If necessary you can subclass HTTP error into individual error classes later. Also, I'd recommend keeping the response json object simple: `{ 'error': 'message' }` should be sufficient.

Ashrjz committed:
Initial commit v0.1

Flask based REST API for v0.1

ashishRaj-0 commented:
I have done the code changes you asked for, let me know if I missed something or if something new needs to be changed.

And where do you like me to add the setup instructions for the API, should I create a `README.md` in `api/` folder or add it in the common `README.md`. I can even add the instructions in the `__init__.py` file if that is fine.

ab-emptycup commented:
Yeah, the code looks much cleaner now. There are a few more points now:

- Keep the route names as `create_pro` instead of `create_profile`. Our convention is that "pro" means "professional". When you are creating a resource or deleting a resource, it is better use a consistent convention. "profile" is a very generic word. When we implement the role for customer (eg: homeowner), you may have another "create_profile" there. Instead naming routes based on the resource name is less ambiguous.
- _utils.py_ should not use _abort_. Utils are generally helper functions. Their job is to try to do an action and if the action fails raise an error. Whether that error should abort the request or not should be decided by the route controller.
- Utils should be independent of _app level_ choices. `read_db()` and `write_db()` now leak the choice that the db is a json file into _utils.py_. `read_db()` and `write_db()` should be helper methods within _pros.py_ (or if necessary _database.py_, but that would be overengineering at this point). _utils.py_ should have `load_json()` and `save_json()`. Those are generic helper methods for loading and saving JSON files. Whether you'll use that json file to save db content or some other data is not relavent for _utils.py_. Think of it like this: tomorrow if you start implementing another API, you should just be able to copy the _utils.py_ there and it should work. _utils.py_ just makes it easy to write code given our code conventions.
- The status messages are missing the Resource IDs for easy tracking. For eg: "No such Professional [%s]" % pid
- Also, another minor detail is capitalising the status messages. Sentence starting and class names should be capitalised. Eg: "No such Professional" is good. "No Such Professional" and "No such professional" are bad.
- Another good practice is to use empty lines to separate logical blocks within a function. For example, let's say your function does 3 logical steps: 1. open files & load data, 2. process that data 3. save the data back into the files. You should have an empty line between those logical steps. That improves code readability significantly. Ofcourse, how to split the function into logical blocks is not an exact science and is more of an art. The hall mark of great code is how well it flows in a reader's mind. A general suggestion is to keep the logical blocks limited to 2-6 lines. 
- Another recommendation is not to `except` all the 500 errors. 500 basically means that the code broke. If the code breaks, you don't need beautiful error reporting. But, you do need beautiful reporting when the request itself is bad. When your code hits a 500, you'll get a traceback in the logs and an opaque error message `500, { 'error': 'Something went wrong. Please try again or contact the admin' }` should be sent to the requester. It's not a good idea to leak info about the system during 500 errors.

I also made some comments in the code. Please do take a look.

ab-emptycup commented on src/api/__init__.py:
```diff
@@ -1 +1,2 @@
-from .app import create_app
\ No newline at end of file
+# For api to be treated as a package, it must contain an __init__.py file 
+# so that blueprint imports can be done in the api.py file.
```
You shouldn't need `__init.py__` for importing from modules within the same folder. May be the dot in `.pros` is the culprit. Try importing directly by doing `from pros import pros`. It should work. If so, you can drop the __init.py__ for now.

ab-emptycup replied:
I think this depends on where you are running the API from.

ab-emptycup commented on src/api/api.py:
```diff
@@ -0,0 +1,34 @@
+
+api = Flask(__name__)
+
+cors.init_app(api)
```
Move initialization of `cors` right about `cors.init_app()`. That way, there's less chance of those getting separated in the future if more code gets added after _api_ initialization.

ab-emptycup commented on src/api/api.py:
```diff
@@ -0,0 +1,34 @@
+
+#error handlers
+@api.errorhandler(HTTPException)
+def http_error(error):
```
Functions should typically be named as verbs. The name should indicate what the function does. `http_error()` < `handle_http_error()`.

ab-emptycup commented on src/api/api.py:
```diff
@@ -0,0 +1,34 @@
+
+
+@api.route('/', methods=['GET'])
+def get_api_routes():
```
`get_routes()` should be good enough here.

ab-emptycup commented on src/api/api.py:
```diff
@@ -0,0 +1,34 @@
+    """
+    Handles all HTTP errors
+    """
+    return jsonify({'error': str(error)}), error.code
```
Inline dictionaries should enclose a space on each side of the brace. For eg: `{ 'error': 'Nuclear explosion' }`.

ab-emptycup reviewed (changes requested):
Made a few suggestions in a comment on the PR and also some inline comments on the latest commit.

ab-emptycup commented:
@ashishRaj-0 Don't worry that this PR is taking time. It's totally cool. Better to pick up these nuances now than later. If necessary, you can push the remaining issues to next week also. Take your time to iron out these little kinks.

Ashrjz committed:
Initial commit v0.1

Flask based REST API for v0.1

ashishRaj-0 commented:
I made suggested changes.
Let me know if any more changes need to be made.

Also, let me know where to add instructions for setup.

ab-emptycup commented on src/api/pros.py:
```diff
@@ -0,0 +1,99 @@
+    """
+    Fetches list of professionals
+    """
+    db = load_json('./src/api/db.json')
```
Generally, any hardcoded value like like path or global constants should be defined in a separate _config.py_. Check out DRY paradigm. It's a good practice.

This need not be changed for now, because we'll be rewriting this part anyway. Just wanted to share for your learning.

ashishRaj-0 replied:
I'll take care of this later, thanks for sharing though.

ab-emptycup commented on src/api/api.py:
```diff
@@ -1 +1,38 @@
+    """
+    routes = { 'routes': [(r.rule, r.endpoint)
+                         for r in api.url_map.iter_rules()] }
+    if not routes:
```
Wouldn't this always fail? Even if `api.url_map.iter_rules()` didn't return anything (which itself is unlikely), _routes_ would still be `{ 'routes': [] }` which would not evaluate to False.

ashishRaj-0 replied:
Yes, I missed that. I took a look at the documentation and I don't think so checking that is required.

ab-emptycup commented on src/api/pros.py:
```diff
@@ -0,0 +1,99 @@
+    """
+    data = request.get_json()
+    if not data:
+        abort(400)
```
Can we have description as to why 400? Eg: `{ 'error': 'Empty request data' }` or something along those lines.

ashishRaj-0 replied:
Done

ab-emptycup commented on src/api/pros.py:
```diff
@@ -0,0 +1,99 @@
+        abort(400)
+
+    uid = uuid.uuid1()
+    if not uid:
```
Can this fail? Documentation doesn't seem to define any failure cases. Did you run into an issue?

ab-emptycup replied:
Also, if you want to just the str identifier, may be you can just do `str(uuid.uuid1()` here.

ashishRaj-0 replied:
Resolved

ab-emptycup commented on src/api/pros.py:
```diff
@@ -0,0 +1,99 @@
+        abort(500)
+    db[str(uid)] = pros
+
+    if save_json('./src/api/db.json', db):
```
It's usually a good pattern to end the function on the success case and escape the errors in if blocks. So, here you would do: `if not save_json(): abort(500)` instead of the other way around.

ab-emptycup replied:
Also, it's better if `save_json` doesn't eat the exception. If it does and just returns _True_ or _False_, the information about why it failed is lost. It may be because there are no permissions or the file already exists or the path doesn't exist etc.

The added advantage is that if `save_json()` throws an Exception, you don't need the `if save_json()` clause here. The exception will automatically get thrown above and you can catch the generic exception just like a HTTPException and return 500 in a separate errorhandler.

ab-emptycup replied:
Anything that is unexpected for you should be 500. Anything that is expected should be a more specific error type. For example, you're trying to open `db.json` and that file should have been created during setup, but it wasn't created. Then it is a 500. 

In a different case, if some file was requested by the request that is not on server, it would be BAD REQUEST. For eg: User's profile image was requested when it was not uploaded. That's not an issue with the API. Both cases the error is FileNotFound but HTTP code is different.

ashishRaj-0 replied:
For now, all the errors raised while opening the `db.json` file will be handled by the calling function of `load_json()` & `save_json` as the errors raised will be `500` error and this is only because a json file is our database, later when a database is incorporated we can generalise these errors into different HTTP exceptions. Do you this makes sense for now?

ab-emptycup commented on src/api/pros.py:
```diff
@@ -0,0 +1,99 @@
+        abort(500)
+
+    if pid in db:
+        return jsonify(db[pid]), 200
```
200 is the default HTTP return code and can be omitted.

ab-emptycup commented on src/api/pros.py:
```diff
@@ -0,0 +1,99 @@
+    db = load_json('./src/api/db.json')
+
+    if pid not in db:
+        return abort(404, description="No such professional {}".format(pid))
```
You can use the `%` operator for brevity.

ab-emptycup commented on src/api/pros.py:
```diff
@@ -0,0 +1,99 @@
+    db[pid]['active'] = datetime.datetime.utcnow().timestamp()
+
+    if save_json('./src/api/db.json', db):
+        return jsonify({ 'message': 'Successfully updated fields {} of user with pid {}'.format([k for k in data.keys()], pid) }), 200
```
It's better to break this line. 

Also, you don't need to do `[k for k in data.keys()]`. You can just do `data.keys()`.

ab-emptycup commented on src/api/utils.py:
```diff
@@ -0,0 +1,32 @@
+import os.path
+
+
+def load_json(file_path):
```
One convention that we follow in our codebase is to end all variable names that are supposed to be paths with `p` for brevity. For eg: "file_path" would be "filep". Similarly any directory path ends with "dirp".

ab-emptycup commented on src/api/utils.py:
```diff
@@ -0,0 +1,32 @@
+
+    with open(file_path, 'r') as rf:
+        data = json.load(rf)
+    if not data:
```
`{}` or `[]` are valid JSON data. They give the info that there is no data which might be valid case.

ashishRaj-0 replied:
Should it return an empty JSON, or a message `No professionals found`

ab-emptycup replied:
It is important to maintain abstractions in the codebase. That means you should not leak information between abstractions. _utils.py_ is a generic module that should usable in some other project as well. That too `load_json()` is just loading a JSON file. What that JSON contains or how it will be interpreted is upto the driver that calls this `utils.load_json()`. If you return "No professionals found", then you're leaking information in the sense that `utils.load_json()` should have no idea what a Professional is.

ashishRaj-0 replied:
For now, all the errors raised while opening the `db.json` file will be handled by the calling function of `load_json()` & `save_json` as the errors raised will be `500` error and this is only because a json file is our database, later when a database is incorporated we can generalise these errors into different HTTP exceptions. Do you this makes sense for now?

ab-emptycup commented on src/api/utils.py:
```diff
@@ -0,0 +1,32 @@
+    Reads from JSON file and returns data.
+    If file not found, creates a new one.
+    """
+    if not os.path.exists(file_path):
```
This is an error. If a helper is asked to read data from a file, the file should exist. The utility should not silently return empty data if the file itself doesn't exist.

ashishRaj-0 replied:
It was done to make the `db.json` available in the first run, but maybe a setup instruction can be done to make it available. Then another case of `FileNotFoundError` can be thrown.

ab-emptycup replied:
I think when you try to open a path "FileNotFoundError" gets thrown automatically, you might not need to throw it explicity.

ashishRaj-0 replied:
Sorry, I mean the calling function of the util function, should it return a `404: No professionals found` for empty database or just return a empty json?

ashishRaj-0 replied:
> This is an error. If a helper is asked to read data from a file, the file should exist. The utility should not silently return empty data if the file itself doesn't exist.

I have removed the part where it was creating a new `db.json` if the file was not found, that can be taken care of while setup, and the error results in a `500` error.

ab-emptycup reviewed (changes requested):
Added a few more suggestions. Do take a look.

ashishRaj-0 commented on src/api/pros.py:
```diff
@@ -0,0 +1,99 @@
+    """
+    db = load_json('./src/api/db.json')
+    if not db:
+        abort(500)
```
Rectified

ab-emptycup commented on src/api/api.py:
```diff
@@ -1 +1,38 @@
+cors.init_app(api)
+
+
+@api.errorhandler(HTTPException)
```
It is a good idea to move this to the end of the file. A good convention for priority is:

1. standard library imports (eg: `import os`)
2. imports from external dependencies (eg: `import flask`)
3. imports from within our codebase (eg: `pros`)
4. module loading code (executed during import itself).
5. public functions (followed by their  private `_helper_func()`s)
6. fallback / catchall functions (followed by their private `_helper_func()` s)


Any code that executes on module loading and not inside a function should typically be at the top of the module (except for driver code `if __name__ == '__main__'` etc, that can be at the bottom). 

The reason is that loading code is kinda like Global context for that module. Anyone seeing that module should know what the context is first. Sometimes, you miss 1 or 2 lines that might be between functions. For eg: `api.register_blueprint(pros)` should be moved to the top

ashishRaj-0 replied:
I removed the part where it creates a new file, if the file is not found, that can be handled during setup and `FileNotFoundError` will be returned by the calling function

ashishRaj-0 replied:
Fixed

Ashrjz committed:
Initial commit v0.1

Flask based RSET API for Portfolio v0.1, with create, read, and update `professionals` list.

ab-emptycup commented on src/api/pros.py:
```diff
@@ -0,0 +1,118 @@
+            'whatsapp': data['whatsapp'] if 'whatsapp' in data else True
+        }
+    except KeyError as e:
+        abort(400, description="%s is missing"%e)
```
A space before and after an operator, eg: `a + b` or `"hello %s" % name`. But not around `=` for keyword arguments.

A few other places in the code as well.

ab-emptycup commented on src/api/utils.py:
```diff
@@ -0,0 +1,19 @@
+def load_json(filep):
+    """
+    Reads from JSON file and returns data.
+    If file not found, creates a new one.
```
I think this is not relavent anymore.

ab-emptycup commented on .gitignore:
```diff
@@ -1,4 +1,7 @@
 build
 data
 portfolio.egg-info
+.vscode/
```
Better to add `*.pyc` also to this list.

Not needed if we do `export PYTHONDONTWRITEBYTECODE=1` to `~/.bashrc`. But, someone working on this might not and might accidently commit a _.pyc_

ab-emptycup reviewed (commented):
Looks pretty good already. Just a couple of light polishes before we merge.

Ashrjz committed:
Initial commit for API v0.1

Flask based REST API for Portfolio v0.1, with create, read,
and update `professionals` list.

ab-emptycup commented:
We can add the instructions in README.md once we add the tables as well. There shouldn't be too many setup instructions just yet.
